acados/ocp_nlp/ocp_nlp_cost_conl.c

```diff
@@ -667,8 +674,22 @@ void ocp_nlp_cost_conl_update_qp_matrices(void *config_, void *dims_, void *mode
         model->conl_cost_fun_jac_hess->evaluate(model->conl_cost_fun_jac_hess, conl_fun_jac_hess_type_in,
                                                 conl_fun_jac_hess_in, conl_fun_jac_hess_type_out, conl_fun_jac_hess_out);
 
-        // hessian of outer loss function
-        blasfeo_dpotrf_l(ny, &work->W, 0, 0, &memory->W_chol, 0, 0);
+        // factorize hessian of outer loss function
+        if (model->psi_hess_is_diag)
+        {
+            printf("sparse factorization\n");
+            double diag_val = 0.;
+            blasfeo_dgese(ny, ny, 0., &memory->W_chol, 0, 0);
+            for (int i = 0; i < ny; i++)
+            {
+                diag_val = sqrt(blasfeo_dgeex1(&work->W, i, i));
+                blasfeo_dgese(1, 1, diag_val, &memory->W_chol, i, i);
```

Comment:
For rather small matrices, this may be slower than just calling the dense Cholesky factorization, since you have two function calls per diagonal element.
I would suggest to perform some benchmark and only switch to the diagonal factorization for sizes where it is beneficial.

I would also suggest to use the macros in here https://github.com/giaf/blasfeo/blob/master/include/blasfeo_common.h#L112 to access the diagonal elements, instead of the two function calls.
The macro works for any BLASFEO version (e.g. is defined differently for each matrix format, e.g. https://github.com/giaf/blasfeo/blob/master/include/blasfeo_common.h#L159 for another one), and it can be used both at the left and right sides of the assignment.

Alternatively, since this is likely a rather common operation, it would be easy to add the corresponding routine directly to BLASFEO, in the auxiliary folder, taking inspiration from existing routines operating on the diagonal of a matrix as e.g. https://github.com/giaf/blasfeo/blob/master/auxiliary/d_aux_lib4.c#L4465 for the panel-major matrix.
A name could be blasfeo_ddiasqrt or something like that.